pytoda/datasets/_csv_statistics.py

```diff
@@ -102,40 +209,45 @@ def reduce_csv_statistics(
     else:
         # sorting the strings
         features = sorted(features)
+
+    # features is the definite features_list
+
     maximums, minimums, means, stds, sample_numbers = zip(
         *[
             (
                 # NOTE: here we ensure that we pick the statistics
-                # in the right order via indexes
+                # in the right order via indexes provided by the dataset
                 csv_dataset.max[csv_dataset.feature_mapping[features].values],
                 csv_dataset.min[csv_dataset.feature_mapping[features].values],
                 csv_dataset.mean[csv_dataset.feature_mapping[features].values],
                 csv_dataset.std[csv_dataset.feature_mapping[features].values],
-                len(csv_dataset)
+                # len(csv_dataset)
+                csv_dataset.notna_count[
+                    csv_dataset.feature_mapping[features].values]
             ) for csv_dataset in csv_datasets
         ]
     )
     # NOTE: reduce max and min
-    maximum = np.array(maximums).max(axis=0)
-    minimum = np.array(minimums).min(axis=0)
+    maximum = np.nanmax(maximums, axis=0)
+    minimum = np.nanmin(minimums, axis=0)
```

Comment:
Are these statistics computed on the imputed or raw values? Imagine a situation where we impute NaNs with 0 (which should be the default), but we only have negative values. Will the maximum be computed correctly here (i.e. 0) or is the imputation applied _after_ the computation of these statistics?

Reply:
The imputation is applied _after_ the transformation on purpose: You know more about the data distribution after the transformation and we can reason on imputation values on these scales. I.e. imputing zero after minmax scaling is setting a minimal value, and for standardisation it is imputing a mean value.

In the case you describe I would think that the goal of minmax scaling would be to map them into the range [0, 1]. Imputing zero beforehand will scew this transformation (lets say the values are in [-2, -1], they would end up in [0, 0.5]) and the imputed zeros will be maximal values (1).

If you know more about your initial data distribution and want impute values that impact the statistics, one would have to impute NaNs explicitly beforehand on the files.

The refactor would also allow you to impute values "manually" using the datasets implementation after it's initialisation, though it requires knowledge of the inner API.

Reply:
Do you think imputing beforehand instead is a common case that should be supported as well?

Reply:
No no, I think imputing afterwards is the only think that makes sense. I just wanted to double check that the code works in this way :)